Mathlib/LinearAlgebra/Goursat.lean

```diff
@@ -0,0 +1,134 @@
+/-
+Copyright (c) 2024 David Loeffler. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: David Loeffler
+-/
+import Mathlib.LinearAlgebra.Prod
+import Mathlib.LinearAlgebra.Quotient.Basic
+import Mathlib.GroupTheory.Goursat
+
+/-!
+# Goursat's lemma for submodules
+
+Let `G, H` be modules over a ring `R`. If `I` is a submodule of `G × H` which projects fully on
+to both factors, then there exist submodules `G' ≤ G` and `H' ≤ H` such that `G' × H' ≤ I` and the
+image of `I` in `(G ⧸ G') × (H ⧸ H')` is the graph of an isomorphism `G ⧸ G' ≃ₗ[R] H ⧸ H'`.
+
+`G'` and `H'` can be explicitly constructed as `Submodule.goursatFst I` and `Submodule.goursatSnd I`
+respectively.
+-/
+
+open Function Set LinearMap
+
+namespace Submodule
+variable {R G H : Type*} [Ring R] [AddCommGroup G] [Module R G] [AddCommGroup H] [Module R H]
+  {I : Submodule R (G × H)}
+  (hI₁ : Surjective (Prod.fst ∘ I.subtype)) (hI₂ : Surjective (Prod.snd ∘ I.subtype))
+
+variable (I) in
+/-- For `I` a submodule of `G × H`, `I.goursatFst` is the kernel of the projection map `I → H`,
+considered as a submodule of `G`.
+
+This is the first subgroup appearing in Goursat's lemma. See `Subgroup.goursat`. -/
+def goursatFst : Submodule R G :=
+  ((LinearMap.snd R G H).comp I.subtype).ker.map ((LinearMap.fst R G H).comp I.subtype)
+
+variable (I) in
+/-- For `I` a subgroup of `G × H`, `I.goursatSnd` is the kernel of the projection map `I → G`,
+considered as a subgroup of `H`.
+
+This is the second subgroup appearing in Goursat's lemma. See `Subgroup.goursat`. -/
+def goursatSnd : Submodule R H :=
+  ((LinearMap.fst R G H).comp I.subtype).ker.map ((LinearMap.snd R G H).comp I.subtype)
+
+lemma goursatFst_toAddSubgroup :
+    (goursatFst I).toAddSubgroup = I.toAddSubgroup.goursatFst := by
+  ext x
+  simp [mem_toAddSubgroup, goursatFst, AddSubgroup.mem_goursatFst]
+
+lemma goursatSnd_toAddSubgroup :
+    (goursatSnd I).toAddSubgroup = I.toAddSubgroup.goursatSnd := by
+  ext x
+  simp [mem_toAddSubgroup, goursatSnd, AddSubgroup.mem_goursatSnd]
```

Comment:
Can't `simps` generate these?

Reply:
I don't think so. The definition of `Submodule.goursatFst` isn't "`AddSubgroup.goursatFst` together with a proof that it's a submodule", it's a new construction a priori independent of `AddSubgroup.goursatFst` and some nonobvious work has to occur to show that the two are actually compatible. So AFAIK this is beyond what `@[simps]` is capable of generating.

Reply:
I am not aware of a way that `@[simps]` can be used to generate these lemmas. If you know of a way of doing so, I would be happy to hear about it.

Reply:
I mean the point of my question was very concretely for you to tag the definition with `simps?` and report on whether it generates the required lemma. If it doesn't then no further action is required

Reply:
No; that was the first thing I tried. `@[simps]` prints a message "not a constructor application, please use `@[simps!]` instead", and `@[simps!]` generates an entirely useless lemma showing that `↑L.goursatFst = (fun a ↦ (↑a).1) '' ↑(ker (LinearMap.snd R M N ∘ₗ L.subtype)) `.

Reply:
I believe there is some syntax for telling `simps` more explicitly what to generate, but I don't know if it is possible to use this to auto-generate `goursatFst_toAddSubgroup`.

Reply:
The syntax would be `simps! toAddSubgroup` but I agree it's unlikely to generate the right lemma

Reply:
Indeed, it doesn't.